File: drills/elixir/solutions/nc92_unique_paths__pascal.exs
defmodule Solution do
  def unique_paths(m, n) when m <= 0 or n <= 0, do: 0

  # Every path is exactly m-1 downs and n-1 rights in some order, so the count
  # is the number of ways to choose which of the m+n-2 moves are downs -- a
  # binomial coefficient, and no grid at all. Multiplying and dividing in step
  # keeps every intermediate an exact integer.
  def unique_paths(m, n) do
    downs = m - 1
    total = m + n - 2

    Enum.reduce(1..downs//1, 1, fn i, acc -> div(acc * (total - downs + i), i) end)
  end
end

File: drills/go/solutions/nc92_unique_paths__pascal.go
package main

func uniquePaths(m int, n int) int {
	// Every path is m-1 downs and n-1 rights in some order: choose which
	// of the m+n-2 steps are downs. Compute the binomial without overflow
	// by multiplying and dividing alternately.
	total, downs := m+n-2, m-1
	result := 1
	for i := 1; i <= downs; i++ {
		result = result * (total - downs + i) / i
	}
	return result
}
